### PR TITLE
[Draft] graph/db: unwrap dns addresses from opaque ones during migration

### PR DESCRIPTION
**Pull request by ellemouton**
This addresses the final TODO in #9795.

Depends on https://github.com/lightningnetwork/lnd/pull/9455

Will clean this up once the dependent PR is merged 👍

### CODE REVIEW

File: graph/db/sql_migration.go [hunk 1]
```diff
@@ -263,14 +263,39 @@ func migrateNodes(ctx context.Context, cfg *sqldb.QueryConfig,
...
+			}
+
+			if byteRead == numAddrBytes {
+				continue
+			}
+
+			addrs = append(addrs, &lnwire.OpaqueAddrs{
+				Payload: opaque.Payload[byteRead:],
+			})
+		}
```
[the remark this answers is not in the document]
Comment (ellemouton) [reply graph/db/sql_migration.go]: will add a comment when im back but did it this way on purpose (test covers it too)

[the remark this answers is not in the document]
Comment (ellemouton) [reply graph/db/sql_migration.go]: ok actually did end up changing it to this

File: graph/db/sql_migration.go [hunk 2]
```diff
@@ -323,12 +322,82 @@ func migrateNodes(ctx context.Context, cfg *sqldb.QueryConfig,
...
+	}
+
+	// Override the node addresses if we have any.
+	if len(addrs) != 0 {
+		node.Addresses = addrs
+	}
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (ellemouton) [reply graph/db/sql_migration.go:399]: @gemini-code-assist
i think that is incorrect. If it does contain emtpy opaque addres, then we want that empty set to remain so that the serialised bytes of the node announcement stays exactly the same otherwise the signature wont be valid

**Comment by ellemouton** [comment]
cc @mohamedawnallah for review 🙏

File: lnwire/lnwire.go [hunk 3]
```diff
@@ -744,146 +744,19 @@ func ReadElement(r io.Reader, element interface{}) error {
...
-				// Now append the rest of the address bytes.
-				_, err := io.ReadFull(addrBuf, payload[1:])
-				if err != nil {
-					return err
-				}
-
-				address = &OpaqueAddrs{
-					Payload: payload,
-				}
-				addrBytesRead = addrsLen
```
Comment (moawnallah) [inline lnwire/lnwire.go:886]: Nice refactoring!

File: lnwire/lnwire.go [hunk 4]
```diff
@@ -949,3 +822,147 @@ func ReadElements(r io.Reader, elements ...interface{}) error {
...
 }
+
+// ReadAddress attempts to read a single address descriptor (as defined in
+// Bolt 7) from the passed io.Reader. The total length of the address section
+// (in bytes) must be provided so that we can ensure we don't read beyond the
+// end of the address section. The number of bytes read from the reader and the
+// parsed net.Addr are returned.
+//
+// NOTE: it is possible for the number of bytes read to be 1 even if a nil
+// address is returned.
```
Comment (moawnallah) [inline lnwire/lnwire.go:833]: 👍

File: lnwire/lnwire.go [hunk 5]
```diff
@@ -949,3 +822,147 @@ func ReadElements(r io.Reader, elements ...interface{}) error {
...
+
+	addrBytesRead := uint16(1)
+
+	var address net.Addr
+	switch aType := addressType(descriptor[0]); aType {
+	case noAddr:
+		return addrBytesRead, nil, nil
+
+	case tcp4Addr:
+		var ip [4]byte
```
Comment (moawnallah) [inline lnwire/lnwire.go:848]: nit: can be `net.IPv4len`

File: lnwire/lnwire.go [hunk 6]
```diff
@@ -949,3 +822,147 @@ func ReadElements(r io.Reader, elements ...interface{}) error {
...
+		}
+
+		address = &net.TCPAddr{
+			IP:   net.IP(ip[:]),
+			Port: int(binary.BigEndian.Uint16(port[:])),
+		}
+		addrBytesRead += tcp4AddrLen
+
+	case tcp6Addr:
+		var ip [16]byte
```
Comment (moawnallah) [inline lnwire/lnwire.go:865]: nit: can be `net.IPv6len`

File: lnwire/lnwire.go [hunk 7]
```diff
@@ -949,3 +822,147 @@ func ReadElements(r io.Reader, elements ...interface{}) error {
...
+		var port [2]byte
+		if _, err := io.ReadFull(addrBuf, port[:]); err != nil {
+			return 0, nil, err
+		}
+
+		address = &DNSAddress{
+			Hostname: string(hostname),
+			Port:     binary.BigEndian.Uint16(port[:]),
+		}
+		addrBytesRead += dnsAddrOverhead + uint16(len(hostname))
```
Comment (moawnallah) [inline lnwire/lnwire.go:943]: 👍

File: graph/db/sql_migration.go [hunk 8]
```diff
@@ -260,18 +260,17 @@ func migrateNodes(ctx context.Context, cfg *sqldb.QueryConfig,
 				"opaque data for node %x: %w", pub, err)
 		}
 
+		if err = maybeOverrideNodeAddresses(node); err != nil {
+			skipped++
+			log.Warnf("Skipping migration of node %x with invalid "+
+				"address (%v): %v", pub, node.Addresses, err)
```
Comment (moawnallah) [inline graph/db/sql_migration.go:266]: Q: Why are we skipping node migration when an invalid DNS address is encountered? Can't we just skip the invalid network address? In the worst-case scenario, the remaining node information is still valid and should be eligible for migration

File: graph/db/sql_migration.go [hunk 9]
```diff
@@ -323,12 +322,82 @@ func migrateNodes(ctx context.Context, cfg *sqldb.QueryConfig,
...
+		// known addresses. We'll do this in a loop, as it's possible
+		// that there are several addresses encoded in a single opaque
+		// address.
+		payload := opaque.Payload
+		for len(payload) > 0 {
+			var (
+				r            = bytes.NewReader(payload)
+				numAddrBytes = uint16(len(payload))
+			)
+			byteRead, readAddr, err := lnwire.ReadAddress(
```
Comment (moawnallah) [inline graph/db/sql_migration.go:361]: nit: `addrRead` instead of `readAddr` similar to `byteRead`

File: graph/db/sql_migration.go [hunk 10]
```diff
@@ -323,12 +322,82 @@ func migrateNodes(ctx context.Context, cfg *sqldb.QueryConfig,
...
+			addrs = append(addrs, addr)
+			continue
+		}
+
+		// For each opaque address, we'll now attempt to parse out any
+		// known addresses. We'll do this in a loop, as it's possible
+		// that there are several addresses encoded in a single opaque
+		// address.
+		payload := opaque.Payload
+		for len(payload) > 0 {
```
Comment (moawnallah) [inline graph/db/sql_migration.go:356]: 👍

File: graph/db/sql_migration.go (hunk 8 above)
Comment (moawnallah) [reply graph/db/sql_migration.go:266]: As I read more the code perhaps invalid DNS address to propagate that violates specification probably worthwhile to skip the entire node migration because of it

File: graph/db/sql_migration_test.go [hunk 11]
```diff
@@ -1174,6 +1246,104 @@ func TestSQLMigrationEdgeCases(t *testing.T) {
 			zombies: []uint64{2},
 		})
 	})
+
+	// We have used this migration as a chance to also extract any DNS
+	// addresses that we previously may have wrapped in an opaque address.
+	// If we do encounter such a case, then the migrated node set will look
+	// slightly different from the original node set in the KV store, and so
+	// we test for that here.
+	t.Run("node with wrapped DNS address inside opaque addr",
```
Comment (moawnallah) [inline graph/db/sql_migration_test.go:1255]: Very nice test and inclusive for many (edge) cases!

File: graph/db/sql_migration.go [hunk 12]
```diff
@@ -323,12 +322,82 @@ func migrateNodes(ctx context.Context, cfg *sqldb.QueryConfig,
...
+			var (
+				r            = bytes.NewReader(payload)
+				numAddrBytes = uint16(len(payload))
+			)
+			byteRead, readAddr, err := lnwire.ReadAddress(
+				r, numAddrBytes,
+			)
+			if err != nil {
+				return err
```
Comment (bhandras) [inline graph/db/sql_migration.go:365]: Does an invalid opaque address payload warrant skipping the entire node from the migration? I think a viable alternative could be just to skip the invalid payload and go on, and if no addresses could be parsed then skip the node altogether. wdyt?

**Review by bhandras — APPROVED** [review]
LGTM 📷

File: lnwire/lnwire.go (hunk 5 above)
Comment (ellemouton) [reply lnwire/lnwire.go:848]: keeping this as a pure code-move commit

File: lnwire/lnwire.go (hunk 6 above)
Comment (ellemouton) [reply lnwire/lnwire.go:865]: keeping this as a pure code-move commit

File: graph/db/sql_migration.go (hunk 8 above)
Comment (ellemouton) [reply graph/db/sql_migration.go:266]: yeah exactly :) i originally had it that way but then decided to just skip the nodes so that we are not propagating invalid nodes. 

Also, this is a far edge case & in all tests i've run, i have not come across any nodes with invalid DNS addrs. So this really is just an edge case. We could really go either way and it wont matter much i think.

File: graph/db/sql_migration.go (hunk 12 above)
Comment (ellemouton) [reply graph/db/sql_migration.go:365]: see reply [here](https://github.com/lightningnetwork/lnd/pull/10162#discussion_r2322348669). lemme know what you think :) 

basically: we cant skip the invalid payload. we either need the entire thing (ie, leave it as opaque addr) or should not migrate the node. otherwise we will not have a valid announcement with a valid sig. 

but yeah, see the other comment re this really being a edge edge case and not something i've actually run into during testing. I dont think it really matters either way. But happy to revert it if yall prefer!

Comment (bhandras) [reply graph/db/sql_migration.go:365]: thanks for the explanation, sgtm! :)
